### PR TITLE
feat: `env` flag for external parameters of docker build and regular build

### PR DESCRIPTION
**Pull request by dj8yfo**
tested docker build cases:

1. no command in manifest via `--env` cli flags [sample_no_workspace](https://github.com/dj8yfo/sample_no_workspace/blob/7a9793727ddd6f8be5482944130a9235f62aa1e9/Cargo.toml) => 

<details>
  <summary>sample-crate-88.testnet</summary><p>
  
```json

{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "--env",
      "GOOGLE_QUERY=https://www.google.com/search?q=google+translate&sca_esv=3c150c50f502bc5d",
      "--env",
      "KEY=VALUE"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-226@sha256:a33c44bc06ae8411458fb23cf57b6d760f51592481276f66f5bc292ed6b80378",
    "contract_path": "",
    "source_code_snapshot": "git+https://github.com/dj8yfo/sample_no_workspace?rev=7a9793727ddd6f8be5482944130a9235f62aa1e9"
  },
  "link": "https://github.com/dj8yfo/sample_no_workspace/tree/7a9793727ddd6f8be5482944130a9235f62aa1e9",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.7.7"
}
```

</p></details>

2.  via `passed_env` in manifest and exported env [sample_no_workspace](https://github.com/dj8yfo/sample_no_workspace/blob/485c69441acc449078b1157b812bb69b744c870d/Cargo.toml)

<details>
  <summary>sample-crate-90.testnet</summary><p>
  
```json

{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "--env",
      "KEY=VALUE",
      "--env",
      "GOOGLE_QUERY=https://www.google.com/search?q=google+translate&sca_esv=3c150c50f502bc5d"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-226@sha256:a33c44bc06ae8411458fb23cf57b6d760f51592481276f66f5bc292ed6b80378",
    "contract_path": "",
    "source_code_snapshot": "git+https://github.com/dj8yfo/sample_no_workspace?rev=485c69441acc449078b1157b812bb69b744c870d"
  },
  "link": "https://github.com/dj8yfo/sample_no_workspace/tree/485c69441acc449078b1157b812bb69b744c870d",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.7.7"
}
```

</p></details>

3. no command in manifest via `--env` cli flags [neardevhub-contract](https://github.com/dj8yfo/neardevhub-contract/blob/bba6a9b63c00dcf5399e1eb2bf8bf18d6d8d2c4e/discussions/Cargo.toml)

<details>
  <summary>discussions.gothamuniquehandle3.community.devhub-test-4.testnet</summary><p>
  
```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "--env",
      "KEY=VALUE",
      "--env",
      "GOOGLE_QUERY=https://www.google.com/search?q=google+translate&sca_esv=3c150c50f502bc5d"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-226@sha256:a33c44bc06ae8411458fb23cf57b6d760f51592481276f66f5bc292ed6b80378",
    "contract_path": "discussions",
    "source_code_snapshot": "git+https://github.com/dj8yfo/neardevhub-contract?rev=bba6a9b63c00dcf5399e1eb2bf8bf18d6d8d2c4e"
  },
  "link": "https://github.com/dj8yfo/neardevhub-contract/tree/bba6a9b63c00dcf5399e1eb2bf8bf18d6d8d2c4e",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.1.0"
}
```
</p></details>

```bash
d673dee7ff8e9e5534796a67154e7d43db27b6e57934c33e2848d6b1b5ebf892  community.devhub-test-4.testnet.wasm
e6ddea282b05a29aae383e46ba7fc26c127caa4de18a8283190a3d86256e70ac  devhub-test-4.testnet.wasm
836a14c755f567a331f9e4ad9f3e99fe696c992525df7c55f8acb18973218c6d  discussions.gothamuniquehandle3.community.devhub-test-4.testnet.wasm
85b3054c02e74e8dbf5eab526e8ee867dce1cc3768aa11b32b4d4110fdc78757  gothamuniquehandle3.community.devhub-test-4.testnet.wasm
85b3054c02e74e8dbf5eab526e8ee867dce1cc3768aa11b32b4d4110fdc78757  standalone-devhub-community-4.testnet.wasm
836a14c755f567a331f9e4ad9f3e99fe696c992525df7c55f8acb18973218c6d  standalone-devhub-discussions-4.testnet.wasm
```

4. via `passed_env` in manifest and exported env ; following cases are deployed from the same source-code [neardevhub](https://github.com/dj8yfo/neardevhub-contract/blob/d05268b3583b8273540d94871f897efa120eb0a5/discussions/Cargo.toml) 
  - no variables exports

    <details>
      <summary>discussions.uniquehandle.community.devhub-no-vars.testnet</summary><p>
  
    ```json
    {
      "build_info": {
        "build_command": [
          "cargo",
          "near",
          "build"
        ],
        "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-226@sha256:a33c44bc06ae8411458fb23cf57b6d760f51592481276f66f5bc292ed6b80378",
        "contract_path": "discussions",
        "source_code_snapshot": "git+https://github.com/dj8yfo/neardevhub-contract?rev=d05268b3583b8273540d94871f897efa120eb0a5"
      },
      "link": "https://github.com/dj8yfo/neardevhub-contract/tree/d05268b3583b8273540d94871f897efa120eb0a5",
      "standards": [
        {
          "standard": "nep330",
          "version": "1.2.0"
        }
      ],
      "version": "0.1.0"
    }
    ```
    </p></details>

    ```bash
    e8fbe501978ac5458a71913ff2fe7ebc1a72573b615ca3fa346f32677d7ca7fc  community.devhub-no-vars.testnet.wasm
    df5992c0f3df6c28b3fa439fdb965cd8ecb8fdec37d65f632affc8de901e0092  devhub-no-vars.testnet.wasm
    4382ba6c1e1e004e8b69d98efc0e7652c2c4cf3c16ae9e2dd4d441e85719523f  discussions.uniquehandle.community.devhub-no-vars.testnet.wasm
    175ae0c5719a21dfab545d418050f91807c858e68983434f4f60ec75cf99f92c  standalone-devhub-community-no-vars.testnet.wasm
    4382ba6c1e1e004e8b69d98efc0e7652c2c4cf3c16ae9e2dd4d441e85719523f  standalone-devhub-discussions-no-vars.testnet.wasm
    175ae0c5719a21dfab545d418050f91807c858e68983434f4f60ec75cf99f92c  uniquehandle.community.devhub-no-vars.testnet.wasm
    ```
  - one variable export 

    <details>
      <summary>discussions.uniquehandle.community.devhub-one-var.testnet</summary><p>
  
    ```json
    {
      "build_info": {
        "build_command": [
          "cargo",
          "near",
          "build",
          "--env",
          "GOOGLE_QUERY=https://www.google.com/search?q=google+translate&sca_esv=3c150c50f502bc5d"
        ],
        "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-226@sha256:a33c44bc06ae8411458fb23cf57b6d760f51592481276f66f5bc292ed6b80378",
        "contract_path": "discussions",
        "source_code_snapshot": "git+https://github.com/dj8yfo/neardevhub-contract?rev=d05268b3583b8273540d94871f897efa120eb0a5"
      },
      "link": "https://github.com/dj8yfo/neardevhub-contract/tree/d05268b3583b8273540d94871f897efa120eb0a5",
      "standards": [
        {
          "standard": "nep330",
          "version": "1.2.0"
        }
      ],
      "version": "0.1.0"
    }
    ```
    </p></details>

    ```bash
    1787219b148d4298ed56cad04ae430c40d92779c6163e6165b88e5207549f41f  community.devhub-one-var.testnet.wasm
    87ecd58d10df0718f599af334601674ffe11f203134358add7b993441a943e19  devhub-one-var.testnet.wasm
    18fc865aaa2990d80175033e93ecd471f954bf269b340d700885c427476d6e9c  discussions.uniquehandle.community.devhub-one-var.testnet.wasm
    ad3b3a4a26618f32c1fde9958dbd4209b7364ff49e20ab3fc515cff6fcb31ca5  standalone-devhub-community-one-var.testnet.wasm
    18fc865aaa2990d80175033e93ecd471f954bf269b340d700885c427476d6e9c  standalone-devhub-discussions-one-var.testnet.wasm
    ad3b3a4a26618f32c1fde9958dbd4209b7364ff49e20ab3fc515cff6fcb31ca5  uniquehandle.community.devhub-one-var.testnet.wasm
    ```

  - two variables export 

    <details>
      <summary>discussions.uniquehandle.community.devhub-two-var.testnet</summary><p>
  
    ```json
    {
      "build_info": {
        "build_command": [
          "cargo",
          "near",
          "build",
          "--env",
          "KEY=VALUE",
          "--env",
          "GOOGLE_QUERY=https://www.google.com/search?q=google+translate&sca_esv=3c150c50f502bc5d"
        ],
        "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-226@sha256:a33c44bc06ae8411458fb23cf57b6d760f51592481276f66f5bc292ed6b80378",
        "contract_path": "discussions",
        "source_code_snapshot": "git+https://github.com/dj8yfo/neardevhub-contract?rev=d05268b3583b8273540d94871f897efa120eb0a5"
      },
      "link": "https://github.com/dj8yfo/neardevhub-contract/tree/d05268b3583b8273540d94871f897efa120eb0a5",
      "standards": [
        {
          "standard": "nep330",
          "version": "1.2.0"
        }
      ],
      "version": "0.1.0"
    }
    ```
    </p></details>

    ```bash
    cda103ad337d322341f2d8ce442c7cac2b57f945f7c37de814c3951f499967f5  community.devhub-two-var.testnet.wasm
    79dc0629f9dde96eb16a823110afd9bcfa6b0e5d7f3b1256fbd3f5da887211b8  devhub-two-var.testnet.wasm
    549d312421e600d583fb9847ddc8142909d6f6c5a75a3a44ef9a1e2198b08443  discussions.uniquehandle.community.devhub-two-var.testnet.wasm
    804550e43cbfe5f463a8b866a4a5eb65ef845cb9b625af93f43a91ec96dd95a3  standalone-devhub-community-two-var.testnet.wasm
    549d312421e600d583fb9847ddc8142909d6f6c5a75a3a44ef9a1e2198b08443  standalone-devhub-discussions-two-var.testnet.wasm
    804550e43cbfe5f463a8b866a4a5eb65ef845cb9b625af93f43a91ec96dd95a3  uniquehandle.community.devhub-two-var.testnet.wasm
    ```

### CODE REVIEW

**Comment by dj8yfo** [comment]
@race-of-sloths ping

**Comment by race-of-sloths** [comment]
@dj8yfo Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!


[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=near/cargo-near/226&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=near/cargo-near/226&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/dj8yfo)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/dj8yfo)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: cargo-near-build/src/types/near/build/input/docker_context.rs [hunk 1]
```diff
@@ -14,100 +14,183 @@ impl super::Opts {
...
+                    "configuring `container_build_command` from cli args, passed to current command".cyan()
+                );
+                println!();
+                Ok(self.passthrough_some_opts_into_docker_cmd())
+            },
+            |mut manifest_command| {
+                self.check_flag_conflicts_with_manifest_command()?;
+                let suffix_env = passed_env.into_iter().flatten().filter(|env_key| {
+                    std::env::var(env_key).is_ok()
+                }).flat_map(|env_key| {
```
Comment (frol) [inline cargo-near-build/src/types/near/build/input/docker_context.rs:37]: nit: It is really hard to reason about. Simple `if let` would be far easier to read, IMO

File: cargo-near-build/src/types/near/build/input/docker_context.rs [hunk 2]
```diff
@@ -14,100 +14,183 @@ impl super::Opts {
...
+        cli_args.extend(self.no_abi.then_some("--no-abi".into()));
+        cli_args.extend(self.no_embed_abi.then_some("--no-embed-abi".into()));
+        cli_args.extend(self.no_doc.then_some("--no-doc".into()));
+
+        cli_args.extend(
+            self.features
+                .clone()
+                .into_iter()
+                .flat_map(|features| ["--features".into(), features]),
+        );
```
Comment (frol) [inline cargo-near-build/src/types/near/build/input/docker_context.rs:84]: nit: I think `if let` would be easier in this case

File: cargo-near-build/src/types/near/build/input/mod.rs [hunk 3]
```diff
@@ -107,6 +111,16 @@ impl Opts {
             color = color_arg.to_string();
             cargo_args.extend(&["--color", &color]);
         }
+
+        let equal_pairs: Vec<String> = self
+            .env
+            .iter()
+            .map(|(key, value)| [key.as_str(), value.as_str()].join("="))
+            .collect();
+        equal_pairs.iter().for_each(|equal_pair| {
```
Comment (frol) [inline cargo-near-build/src/types/near/build/input/mod.rs:120]: nit: I find simple `for` loop easier to read than `iter().for_each() + closure`

**Review by frol — CHANGES_REQUESTED** [review]
Feel free to ignore `nit` comments, but there is one that is important

File: cargo-near/src/commands/new/new-project-template/Cargo.toml.template [hunk 4]
```diff
@@ -15,9 +15,14 @@ crate-type = ["cdylib", "rlib"]
...
-image = "sourcescan/cargo-near:0.8.2-rust-1.81.0"
+# TODO: replace with git-image tag after publish 
+image = "sourcescan/cargo-near:git-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-1.81.0"
 # tag after colon above serves only descriptive purpose; image is identified by digest
-image_digest = "sha256:3b5e6efa1d0cef710009f9cc20c845d316187cc170455edc5ba7fdf471c19655"
+image_digest = "sha256:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+# list of environment variables names, whose values, if set, will be used as external build parameters
+# in a reproducible manner
+# supported by `git-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-1.81.0` image or later images
+passed_env = []
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply cargo-near/src/commands/new/new-project-template/Cargo.toml.template]: resolved by https://github.com/near/cargo-near/pull/226/commits/99c3be15af939c61bff63f833e6ba82b1af7c0ef

File: cargo-near-build/src/near/build/mod.rs [hunk 5]
```diff
@@ -50,7 +50,15 @@ pub fn run(args: Opts) -> eyre::Result<CompilationArtifact> {
...
-    let mut build_env = vec![("RUSTFLAGS", "-C link-arg=-s")];
+    let mut build_env = {
+        let mut build_env = vec![("RUSTFLAGS", "-C link-arg=-s")];
+        build_env.extend(
+            args.env
+                .iter()
+                .map(|(key, value)| (key.as_ref(), value.as_ref())),
+        );
+        build_env
+    };
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply cargo-near-build/src/near/build/mod.rs]: removed nesting , but `concat` would require additional `collect`, and there's one more extend, that will be needed in one of the following prs

File: cargo-near-build/src/near/docker_build/subprocess_step.rs [hunk 6]
```diff
@@ -46,9 +46,16 @@ pub fn run(
...
         let env_args = env.docker_args();
-        let cargo_cmd = opts
-            .get_cli_build_command_in_docker(docker_build_meta.container_build_command.clone())?;
+        let cargo_cmd = {
+            let cargo_cmd = opts.get_cli_build_command_in_docker(
+                docker_build_meta.container_build_command.clone(),
+                docker_build_meta.passed_env.clone(),
+            )?;
+            tracing::debug!("cli_build_command_in_docker {:#?}", cargo_cmd);
+            shell_words::join(cargo_cmd)
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply cargo-near-build/src/near/docker_build/subprocess_step.rs:55]: renamed to `shell_escaped_cargo_cmd` . not possible to move to 84 due to blocks nesting and borrowing of `String` to `&str`

File: cargo-near-build/src/types/near/build/input/docker_context.rs [hunk 7]
```diff
@@ -14,100 +14,183 @@ impl super::Opts {
...
-            }
-            if self.no_default_features {
-                return Err(eyre::eyre!(format!(
-                    "`{}` {}",
-                    "--no-default-features",
-                    Self::BUILD_COMMAND_CLI_CONFIG_ERR
-                )));
+        passed_env: Option<Vec<String>>,
+    ) -> eyre::Result<Vec<String>> {
+        manifest_command.map_or_else(
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply cargo-near-build/src/types/near/build/input/docker_context.rs]: i beleive the size is the same, but `if let Some()` is met more often and therefore is easier to understand

**Review by frol — APPROVED** [review]
@race-of-sloths score 8

**Pull request merged**